### PR TITLE
【OSPP】Json check compatiblity

### PR DESCRIPTION
**Pull request by Tomoko-hjf**
## What is the purpose of the change
【OSPP】Json check compatiblity

## Brief changelog
* Check each interface for json compatibility
* Set `checkJsonCompatibility` attribute of `ProtocolConfig` to enable compatibility check. Compatibility check is enabled by default

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java [hunk 1]
```diff
@@ -113,6 +122,15 @@ public void afterUnExport() {
         return exporter;
     }
 
+    private void checkJsonCompatibility(Class<?> clazz) throws RpcException {
+        boolean compatibility = JsonCompatibilityUtil.checkClassCompatibility(clazz);
+        if (!compatibility) {
+            List<String> unsupportedMethods = JsonCompatibilityUtil.getUnsupportedMethods(clazz);
+            logger.error("Interface {} does not support json serialization, the specific methods are {}", clazz, unsupportedMethods);
+            throw new RpcException(String.format("Interface %s does not support json serialization", clazz));
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java:130]: Print logs or affect startup, should be decided by the user, it should be a configuration, configuration json check levels can be ignore (no check), log, startup error, respectively can correspond to close, alarm, strict

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java (hunk 1 above)
Comment (Tomoko-hjf) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java]: done!
In the new version, user can specify the implementation mode by specifying attribute `json-check-level`.  The optional values are `ignore`, `log`, or `startup-error`. 
If no value is specified, the default implementation mode is `startup-error`.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java [hunk 2]
```diff
@@ -113,6 +122,31 @@ public void afterUnExport() {
...
+            boolean compatibility = JsonCompatibilityUtil.checkClassCompatibility(clazz);
+            if (!compatibility) {
+                List<String> unsupportedMethods = JsonCompatibilityUtil.getUnsupportedMethods(clazz);
+                assert unsupportedMethods != null;
+                throw new RpcException(String.format("Interface %s does not support json serialization, the specific methods are %s.", clazz.getName(), unsupportedMethods));
+            } else {
+                logger.info("Check json compatibility complete, all methods can be serialized using json.");
+            }
+        } else {
+            logger.info("Ignore json compatibility check.");
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java:146]: No need to print.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java [hunk 3]
```diff
@@ -113,6 +122,31 @@ public void afterUnExport() {
...
+    private void checkJsonCompatibility(Class<?> clazz, String jsonCheckLevel) throws RpcException {
+
+        if (jsonCheckLevel == null || JSON_CHECK_LEVEL_WARN.equals(jsonCheckLevel)) {
+            boolean compatibility = JsonCompatibilityUtil.checkClassCompatibility(clazz);
+            if (!compatibility) {
+                List<String> unsupportedMethods = JsonCompatibilityUtil.getUnsupportedMethods(clazz);
+                assert unsupportedMethods != null;
+                logger.warn("", "", "", String.format("Interface %s does not support json serialization, the specific methods are %s.", clazz.getName(), unsupportedMethods));
+            } else {
+                logger.info("Check json compatibility complete, all methods can be serialized using json.");
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java:134]: The log level should be adjusted to debug, and the check class should be added to the log message.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java [hunk 4]
```diff
@@ -113,6 +122,31 @@ public void afterUnExport() {
...
+            }
+        } else if (JSON_CHECK_LEVEL_STRICT.equals(jsonCheckLevel)) {
+            boolean compatibility = JsonCompatibilityUtil.checkClassCompatibility(clazz);
+            if (!compatibility) {
+                List<String> unsupportedMethods = JsonCompatibilityUtil.getUnsupportedMethods(clazz);
+                assert unsupportedMethods != null;
+                throw new RpcException(String.format("Interface %s does not support json serialization, the specific methods are %s.", clazz.getName(), unsupportedMethods));
+            } else {
+                logger.info("Check json compatibility complete, all methods can be serialized using json.");
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java:143]: As above.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/constans/RestConstant.java [hunk 5]
```diff
@@ -57,6 +57,11 @@ public interface RestConstant {
     String IDLE_TIMEOUT_PARAM = "idle.timeout";
     String KEEP_ALIVE_TIMEOUT_PARAM = "keep.alive.timeout";
 
+    String JSON_CHECK_LEVEL = "jsonCheckLevel";
+    String JSON_CHECK_LEVEL_DISABLED = "disable";
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/constans/RestConstant.java:61]: change to `disabled`

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java (hunk 2 above)
Comment (Tomoko-hjf) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java]: done

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java (hunk 4 above)
Comment (Tomoko-hjf) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java]: done

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/constans/RestConstant.java (hunk 5 above)
Comment (Tomoko-hjf) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/constans/RestConstant.java]: done

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java (hunk 3 above)
Comment (Tomoko-hjf) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java]: done

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java [hunk 6]
```diff
@@ -113,6 +122,29 @@ public void afterUnExport() {
...
+                logger.warn("", "", "", String.format("Interface %s does not support json serialization, the specific methods are %s.", clazz.getName(), unsupportedMethods));
+            } else {
+                logger.debug("Check json compatibility complete, all methods of {} can be serialized using json.", clazz.getName());
+            }
+        } else if (JSON_CHECK_LEVEL_STRICT.equals(jsonCheckLevel)) {
+            boolean compatibility = JsonCompatibilityUtil.checkClassCompatibility(clazz);
+            if (!compatibility) {
+                List<String> unsupportedMethods = JsonCompatibilityUtil.getUnsupportedMethods(clazz);
+                assert unsupportedMethods != null;
+                throw new RpcException(String.format("Interface %s does not support json serialization, the specific methods are %s.", clazz.getName(), unsupportedMethods));
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java:141]: change `RpcException` to `IllegalStateException`.

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java (hunk 6 above)
Comment (Tomoko-hjf) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java]: done

**Review by CrazyHZM — APPROVED** [review]

**Pull request merged**